test/unit/test_ranges.py

```diff
@@ -0,0 +1,223 @@
+        self.assertEqual(r.status, 206)
+        response_headers = dict(r.getheaders())
+        r.read()
+        self.assertEqual(response_headers['content-range'], 'bytes 95-99/100')
+
+        # Get first 5 bytes
+        headers = {'RANGE': 'bytes=0-4'}
+        self.expected_headers = headers
+        proxy_conn.request('GET', object_path, headers=headers)
+        r = proxy_conn.getresponse()
+        self.assertEqual(r.status, 206)
+        response_headers = dict(r.getheaders())
+        r.read()
+        self.assertEqual(response_headers['content-range'], 'bytes 0-4/100')
+
+        # Get 20 intermediate bytes
+        headers = {'RANGE': 'bytes=20-40'}
+        self.expected_headers = headers
+        proxy_conn.request('GET', object_path, headers=headers)
+        r = proxy_conn.getresponse()
+        self.assertEqual(r.status, 206)
+        response_headers = dict(r.getheaders())
+        r.read()
+        self.assertEqual(response_headers['content-range'], 'bytes 20-40/100')
+
+    def validate_request_headers(self, method, path, headers):
+        """
+        Validate headers of requests for an entity matching `check_path`
+        with method `check_method`.
+        """
+        if (self.check_method == method and self.check_path == path):
+            for header in self.expected_headers:
+                self.assertEqual(headers[header],
+                                 self.expected_headers[header])
+
+    def setup_ring(self, path, **kwargs):
+        """
+        Setup a single replica ring.
+        """
+        dev = {
+            'id': 0,
+            'zone': 0,
+            'device': 'sda1',
+            'ip': '127.0.0.1',
+            'port': 6000
+        }
+        dev.update(kwargs)
+
+        # Prepare ring directory
+        directory, filename = os.path.split(path)
+        try:
+            os.makedirs(os.path.join(directory, dev['device']))
+        except OSError as e:
+            # Ignore if directory already exists
+            if e.errno != 17:
+                raise
+
+        # Compose ring metadata
+        replica2part2dev = [[0, 0, 0, 0]]
+        part_shift = 30
+        ring = RingData(replica2part2dev, [dev], part_shift)
+        with closing(GzipFile(path, 'wb')) as f:
+            pickle.dump(ring, f)
+
+    def sproxyd_mock(self, env, start_response):
+        method = env['REQUEST_METHOD']
+        path = env['PATH_INFO']
+
+        if method == 'GET' and path == "%s/.conf" % self.sproxyd_path:
+            # Return configuration parameter expected by failure detector
+            response = json.dumps({
+                'by_path_enabled': True,
+            }, indent=2)
+        else:
+            response = self.response
+
+        # Expected metadata (xattr) headers by swift
+        metadata = {
+            'X-Timestamp': normalize_timestamp(time.time()),
+            'Content-Length': len(response),
+            'ETag': "some_hash",
+        }
+
+        request_headers = dict([
```

Comment:
I'd use slightly different newline placement (and no need for list comprehension):
```python
request_headers = dict(
    (k[len('HTTP_':], v)
    for k, v in env.iteritems()
    if k.startswith('HTTP_'))
```